### ISSUE TITLE
False positive triggered by cgo pre-processed code in the go build cache

### ISSUE THREAD

**Issue by jvmatl** [dominikh/go-tools#673]
I have a code repo that includes some CGO code. Because of a quirk in how we build within docker containers, the go cache directory is inside the git repo. Staticcheck appears to enter the go build cache and analyze the pre-processed go code in there as well - which would be fine, except for one file whose cached, preprocessed version generates an error:

The error:
`.cache/alpine-gobe/go-build/32/32a39fbab36fd1401a89d7b34311dbbac9d71940b237c56de6ae7b584dc7f729-d:3:1: package comment should be of the form "Package akasim ..." (ST1000)`

The top 11 lines of the offending file:
```
// Code generated by cmd/cgo; DO NOT EDIT.

//line /go/code/pkg/akasim/client.go:1:1
// Package akasim wraps client-side AKA functions to simulate the aka hardware
package akasim; import _cgo_unsafe "unsafe"

/*
#cgo CFLAGS: -I ${SRCDIR}/cgo/include -DOPENSSL
#cgo LDFLAGS: -L ${SRCDIR}/cgo/lib -laka -lcrypto
*/
```

the top of the __original__ source file (before cgo processing) looks like:
```
// Package akasim wraps client-side AKA functions to simulate the aka hardware
package akasim

/*
#cgo CFLAGS: -I ${SRCDIR}/cgo/include -DOPENSSL
#cgo LDFLAGS: -L ${SRCDIR}/cgo/lib -laka -lcrypto
*/
```

It appears the ST1000 is being triggered because the package comment returned by f.Doc.Text() (called in `stylecheck/lint.go:CheckPackageComment` line 51 or so, begins with `line /go/code...` instead of `Package akasim ...`

This is, admittedly, a quirky build setup, and I was able to work around it by moving the package comment to another file in the package that does not contain any cgo, but I thought you might find it interesting, since other false positives might get triggered the same way. If this configuration is one you decide you don't care to worry about, I totally understand. : )

Obviously, one option would be to skip any processing of files in the build cache. 

Another would be to try to detect and handle this specific problem. Since you have no control over the cgo processing, the only thing you could really do here is put in a heuristic to skip comments that match `^//line go:[0-9]+.[0-9]+$` or somesuch (depending on how finicky you want your regex) 

- The output of 'staticcheck -version'

`staticcheck 2019.2.3`

- The output of 'staticcheck -debug.version' 

```
staticcheck 2019.2.3

Compiled with Go version: go1.13.6
Main module:
        honnef.co/go/tools@v0.0.1-2019.2.3 (sum: h1:3JgtbtFHMiCmsznwGVTUWbgGov+pVqnlf1dEJTNAXeM=)
Dependencies:
        github.com/BurntSushi/toml@v0.3.1 (sum: h1:WXkYYl6Yr3qBf1K79EBnL4mak0OimBfB0XUf9Vl28OQ=)
        golang.org/x/tools@v0.0.0-20200116181651-872a348c3885 (sum: h1:ord1pqXhKRYcujhzmNg23POL8yEINk3Mj3fqWS/SNG4=)
```

- The output of 'go version'

`go version go1.13.6 linux/amd64`

- The output of 'go env'

```
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/go/code/.cache/alpine-gobe/go-build"
GOENV="/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/go/code/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build905760856=/tmp/go-build -gno-record-gcc-switches"
```

- Exactly which command you ran

`staticcheck ./...` (from `/go/code`)

- Output of the command and what's wrong with the output

`see above`

- Where we can read the code you're running staticcheck on
  (GitHub repo, link to playground, code embedded in the issue, ...)

`see above`

**Comment by dominikh** [comment]
Note that where your build cache is located at isn't actually relevant for this false positive. It would occur even with a build cache located outside of the repository.

When checking packages that use cgo, we're really checking the output of the cgo preprocessor (as before preprocessing, the code isn't technically valid Go). In the normal case, when we find an issue in a generated file, we report the error in the original file – that's what the `//line` directives are for: so that the compiler and tools can report useful errors that don't point into the preprocessed files.

The reason we're reporting the error in the generated file, by the looks of it, is that the error technically occurs "before" (or rather exactly on) the first `//line` directive, so we don't have a valid mapping to the source file yet. We should probably look into fixing that. Ideally we never report errors in the build cache, and always in the source file.

The rest of your analysis is correct. Go places the `//line` directive in an unfortunate manner, and we should work around that in ST1000.

**Comment by dominikh** [comment]
I believe this was fixed by golang/go@5a550b695117f07a4f2454039a4871250cd3ed09.
